src/cc/arduino/plugins/wifi101/flashers/java/NinaFlasher.java

```diff
@@ -0,0 +1,139 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As a special exception, you may use this file as part of a free software
+ * library without restriction.  Specifically, if other files instantiate
+ * templates or use macros or inline functions from this file, or you compile
+ * this file and link it with other files to produce an executable, this
+ * file does not by itself cause the resulting executable to be covered by
+ * the GNU General Public License.  This exception does not however
+ * invalidate any other reasons why the executable file might be covered by
+ * the GNU General Public License.
+ */
+package cc.arduino.plugins.wifi101.flashers.java;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.*;
+import java.util.List;
+import java.io.ByteArrayOutputStream;
+import java.io.*;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import org.apache.commons.lang3.StringUtils;
+import java.security.MessageDigest;
+
+import cc.arduino.plugins.wifi101.certs.WiFi101Certificate;
+import cc.arduino.plugins.wifi101.certs.WiFi101CertificateBundle;
+
+import cc.arduino.plugins.wifi101.flashers.Flasher;
+import javax.swing.JProgressBar;
+
+public class NinaFlasher extends Flasher {
+
+	public byte[] md5Checksum;
+
+	public NinaFlasher(String _modulename, String _version, String _filename, boolean _certavail, ArrayList<String> _compatibleBoard) {
+		super(_modulename, _version, _filename, _certavail, _compatibleBoard);
+	}
+
+	@Override
+	public void updateFirmware(String port) throws Exception {
+		FlasherSerialClient client = null;
+		try {
+			file = openFirmwareFile();
+			progress(10, "Connecting to programmer...");
+			client = new FlasherSerialClient();
+			client.open(port);
+			client.hello();
+			int maxPayload = client.getMaximumPayload();
+
+			byte[] fwData = this.getData();
+			int size = fwData.length;
+			int address = 0x00000000;
+			int written = 0;
+
+			progress(20, "Erasing target...");
+			client.eraseFlash(address, size);
+
+			while (written < size) {
+				progress(20 + written * 40 / size, "Programming " + size + " bytes ...");
+				int len = maxPayload;
+				if (written + len > size) {
+					len = size - written;
+				}
+				client.writeFlash(address, Arrays.copyOfRange(fwData, written, written + len));
+				written += len;
+				address += len;
+			}
+
+			progress(55, "Verifying...");
+
+			String filepath = file.getPath();
+			md5Checksum = getMD5Checksum(filepath);
+
+			address = 0x00000000;
+			byte[] md5rec = client.md5Flash(address, size);
+			progress(75, "Verifying...");
+			if (Arrays.equals(md5rec, md5Checksum)) {
+				progress(100, "Done!");
+			} else {
+				throw new Exception("Error different MD5 Checksum.");
```

Comment:
Maybe a nicer error message here? Something like: "Error validating flashed firmware"

Reply:
i had change but i loose it, i re-modify and push it again